node/src/components/network/conman.rs

```diff
@@ -0,0 +1,1054 @@
+        let mut outgoing_handler = {
+            let mut guard = ctx.state.write().expect("lock poisoned");
+
+            if guard.address_book.contains(&peer_addr) {
+                debug!("got raced by another outgoing handler, aborting");
+                return;
+            }
+
+            let now = Instant::now();
+            if guard.should_not_call(&peer_addr, now) {
+                // This should happen very rarely, it requires a racing handler to complete and the
+                // resulting do-not-call to expire all while this function was starting.
+                debug!("address turned do-not-call");
+                return;
+            }
+            guard.prune_should_not_call(&peer_addr);
+
+            Self::new(&mut *guard, ctx.clone(), peer_addr)
+        };
+
+        // We now enter a connection loop. After attempting to connect and serve, we either sleep
+        // and repeat the loop, connecting again, or `break` with a do-not-call timer.
+        let do_not_call_until = loop {
+            // We need a subspan to avoid duplicate registrations of peer data on retries.
+            let sub_span = error_span!("connect-and-serve", peer_id = Empty, consensus_key = Empty);
+            match outgoing_handler
+                .connect_and_serve()
+                .instrument(sub_span)
+                .await
+            {
+                Ok(duration) => {
+                    // Regular connection closure, i.e. without an error reported.
+
+                    // Judge how long the connection was active.
+                    let delay = if duration > ctx.cfg.flaky_connection_threshold {
+                        rate_limited!(LOST_CONNECTION, |dropped| info!(
+                            dropped,
+                            "lost connection, will reconnect"
+                        ));
+                        ctx.cfg.successful_reconnect_delay
+                    } else {
+                        rate_limited!(LOST_FLAKY_CONNECTION, |dropped| warn!(
+                            dropped,
+                            "lost connection, but its flaky, will reconnect later"
+                        ));
+                        ctx.cfg.significant_error_backoff
+                    };
+
+                    tokio::time::sleep(delay).await;
+
+                    // After this, the loop will repeat, triggering a reconnect.
+                }
+                Err(OutgoingError::EncounteredBannedPeer(until)) => {
+                    // We will not keep attempting to connect to banned peers, put them on the
+                    // do-not-call list.
+                    break until;
+                }
+                Err(OutgoingError::FailedToCompleteHandshake(err)) => {
+                    debug!(%err, "failed to complete handshake");
+                    break Instant::now() + ctx.cfg.significant_error_backoff;
+                }
+                Err(OutgoingError::LoopbackEncountered) => {
```

Comment:
This should never happen, right? We should never be asked to connect to loopback. Shouldn't we return in this case?

Reply:
Well the point is somewhat moot, since yes, this should actually be `unreachable!`. Getting to this point is a serious bug. However, the current version `returns` here, so all good.